src/src/PluginStructs/P122_data_struct.cpp

```diff
@@ -0,0 +1,568 @@
+      {
+        _state  = P122_state::Wait_for_reset;  // Will take <15ms according to datasheet
+        _last_action_started = millis();
+      }
+    break;
+
+    case P122_state::Wait_for_reset:
+      if (timeOutReached(_last_action_started + P122_RESET_DELAY)) //we need to wait for the chip to reset
+      {
+        if (I2C_wakeup(_i2caddr) != 0)
+        {
+          _errCount++;
+          _state = P122_state::Uninitialized;  // Retry
+        }
+        else
+        {
+          _errCount = 0;   // Device is reachable and initialized, reset error counter
+          _state = P122_state::Read_eid;
+        }
+      }
+    break;      
+
+    case P122_state::Read_eid:
+      _eida = getEIDA();
+      _eidb = getEIDB();
+      _firmware = getFirmwareVersion();
+      _state = P122_state::Write_user_reg;
+    break;
+
+    case P122_state::Write_user_reg:
+      writeUserReg();
+      _state = P122_state::Initialized;
+    break;
+
+    case P122_state::Initialized:
+      // For now trigger the first read cycle automatically
+      _state = P122_state::Ready;
+    break;
+
+    case P122_state::Ready:
+      // Ready to execute a measurement cycle
+      // Start measuring temperature
+      if (!I2C_write8(_i2caddr, SHT2x_GET_TEMPERATURE_NO_HOLD))
+      {
+        _errCount++;
+        _state = P122_state::Uninitialized;  // Retry
+      }
+      else
+      {
+        _last_action_started = millis();
+        _state = P122_state::Wait_for_temperature_samples;
+      }
+    break;
+
+   case P122_state::Wait_for_temperature_samples:
+      if (timeOutReached(_last_action_started + getTempDuration())) 
+      {
+        if (!readTemperature()){    // Read the previously measured temperature
+          _errCount++;
+          _state = P122_state::Uninitialized; // Lost connection
+        }
+        else if (!requestHumidity())    // Start humidity measurement
+        {
+          _errCount++;
+          _state = P122_state::Uninitialized; // Lost connection
+        }
+        else
+        {
+          _last_action_started = millis();
+          _state = P122_state::Wait_for_humidity_samples;
+        }
+      }
+    break;
+
+    case P122_state::Wait_for_humidity_samples:
+      //make sure we wait for the measurement to complete
+      if (timeOutReached(_last_action_started + getHumDuration())) 
+      {
+        if (!readHumidity()) 
+        {
+          _errCount++;
+          _state = P122_state::Uninitialized; // Lost connection
+        }
+        else
+        {
+          _last_action_started = millis();
+          _state = P122_state::New_Values_Available;
+          stable = true;
+        }
+      }
+    break;
+
+    case P122_state::Error:
+    case P122_state::New_Values_Available:
+      //this state is used outside so all we need is to stay here
+      stable = true;
+    break;
+
+    default: //anything not defined above sends the device for initialization
```

Comment:
Having a `default` case in a switch statement that uses an `enum class` type variable defeats the advantage of having the `enum class`, as the compiler now _can't_ check if all values of the enum are handled in the switch. (this check is enabled in ESPEasy).

Reply:
I didn't know about this check. Having a FSM running with an undefined state is a disater I wanted to prevent.